Sources/NativeBuilder/ContainerBuildIR/Graph/GraphBuilder+Arg.swift

```diff
@@ -0,0 +1,99 @@
+    @discardableResult
+    public func arg(_ name: String, defaultValue: String? = nil) throws -> Self {
+        let operation = MetadataOperation(
+            action: .declareArg(name: name, defaultValue: defaultValue)
+        )
+        return try add(operation)
+    }
+
+    /// Add a FROM-only build argument (ARG before the first FROM).
+    /// These arguments are only available in FROM instructions, not within stages.
+    @discardableResult
+    public func fromOnlyArg(_ name: String, defaultValue: String? = nil) -> Self {
+        buildArgs[name] = defaultValue
+        return self
+    }
+
+    /// Check if there's an active stage.
+    public var hasActiveStage: Bool {
+        currentStage != nil
+    }
+
+    /// Resolve an ARG value.
+    /// - Parameters:
+    ///   - key: The ARG name to resolve.
+    ///   - inFromContext: Whether this is being called from a FROM instruction context.
+    /// - Returns: The resolved ARG value, or nil if not found.
+    public func resolveArg(key: String, inFromContext: Bool = false) -> String? {
+        // A global FROM-only ARG
+        guard let currentStage, !inFromContext else {
+            return buildArgs[key]
+        }
+
+        let (found, defaultValue) = currentStage.getDeclaredArg(key)
+        if found {
+            // A stage-local ARG with a default value
+            if let defaultValue {
+                return defaultValue
+            }
+            // A stage-local ARG without a default value - a redeclared global FROM-only ARG
+            return buildArgs[key]
+        }
+
+        return nil
+    }
+
+    /// Substitute ARG variables in a string.
+    /// - Parameters:
+    ///   - input: The string that may contain `${ARG}` references.
+    ///   - inFromContext: Whether this substitution is happening in a FROM instruction context.
+    /// - Returns: The string with ARG variables substituted.
+    public func substituteArgs(_ input: String, inFromContext: Bool) -> String {
+        let pattern = #"\$\{([A-Za-z_][A-Za-z0-9_]*)\}"#  // `${ARG}`
+        let regex = try! NSRegularExpression(pattern: pattern)
+        let range = NSRange(location: 0, length: input.count)
+
+        var result = input
+        var offset = 0
+
+        regex.enumerateMatches(in: input, range: range) { match, _, _ in
+            guard let match, let varRange = Range(match.range(at: 1), in: input) else {
+                return
+            }
+
+            let varName = String(input[varRange])
+            let replacement = resolveArg(key: varName, inFromContext: inFromContext) ?? ""
+
+            let matchRange = match.range
+            let adjustedMatchRange = NSRange(location: matchRange.location + offset, length: matchRange.length)
+
+            if let replacementRange = Range(adjustedMatchRange, in: result) {
+                result.replaceSubrange(replacementRange, with: replacement)
+                offset += replacement.count - matchRange.length
```

Comment:
`String.count` and `Range.length` work on different principles

`String.count`  counts as grpaheme clusters (i.e. Ç would be 1 count)
`Range.length` counts as utf-16 (i.e Ç would be two counts)